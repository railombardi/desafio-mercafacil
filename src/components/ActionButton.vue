<template>
  <label class="wrapper">
    <component class="icon" v-if="icon" :is="iconComponent" />
    <button class="button">
      {{ label }}
    </button>
  </label>
</template>

<script>
export default {
  name: "ActionButton",
  props: {
    label: {
      type: String,
      default: "label",
    },
    icon: {
      type: String,
      default: "",
    },
  },
  computed: {
    iconComponent() {
      return `${this.icon}-icon`;
    },
  },
};
</script>

<style lang="scss" scoped>
.wrapper {
  width: 100%;
  position: relative;
  .icon {
    position: absolute;
    top: calc(50% - 12px / 2);
    left: 22px;
  }
  .button {
    width: 100%;
    height: 56px;
    background: $primary-50;
    box-shadow: 0px 0px 2px rgba(0, 0, 0, 0.12), 0px 2px 2px rgba(0, 0, 0, 0.24);
    border-radius: 4px;
    padding: 10px 16px;
    color: $primary-500;
    font-weight: 500;
    font-size: 14px;
    line-height: 16px;
    border: none;
  }
}
</style>
