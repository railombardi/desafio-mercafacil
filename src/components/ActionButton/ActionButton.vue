<template>
  <label class="wrapper">
    <component class="icon" v-if="icon" :is="iconComponent" />
    <button @click="emitClick" class="button">
      {{ label }}
    </button>
  </label>
</template>

<script>
export default {
  name: "ActionButton",
  props: {
    /**
     * button label
     */
    label: {
      type: String,
      default: "label",
    },
    /**
     * intended if you want to use any icon from the icon components
     */
    icon: {
      type: String,
      default: "",
    },
  },
  computed: {
    iconComponent() {
      return `${this.icon}-icon`;
    },
  },
  methods: {
    emitClick(e) {
      /**
       * Passthrough click event
       * @type {Event}
       */
      this.$emit("click", e);
    },
  },
};
</script>

<style lang="scss" scoped>
.wrapper {
  width: 100%;
  position: relative;
  .icon {
    position: absolute;
    top: calc(50% - 12px / 2);
    left: 22px;
  }
  .button {
    width: 100%;
    height: 56px;
    background: $primary-50;
    box-shadow: 0px 0px 2px rgba(0, 0, 0, 0.12), 0px 2px 2px rgba(0, 0, 0, 0.24);
    border-radius: 4px;
    padding: 10px 16px;
    color: $primary-500;
    font-weight: 500;
    font-size: 14px;
    line-height: 16px;
    border: none;
    cursor: pointer;
  }
}
</style>
