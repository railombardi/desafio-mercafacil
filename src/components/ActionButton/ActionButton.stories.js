import ActionButton from "./ActionButton.vue";
import MenuIcon from "../icons/MenuIcon.vue";
import SearchIcon from "../icons/SearchIcon.vue";
import FilterIcon from "../icons/FilterIcon.vue";
export default {
  title: "Components/ActionButton",
  component: ActionButton,
  argTypes: {
    icon: {
      control: { type: "select" },
      options: ["menu", "filter", "search"],
    },
  },
};

const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: {
    ActionButton,
    MenuIcon,
    SearchIcon,
    FilterIcon,
  },
  template: `<div>
              <action-button @click.native="onClick" :label="label" :icon="icon" />
              <span style="margin-top: 15px;"> Clicked count: {{clickedCount}} </span>
             </div>`,
  data() {
    return {
      clickedCount: 0,
    };
  },
  methods: {
    onClick() {
      this.clickedCount += 1;
    },
  },
});

export const Default = Template.bind({});
Default.args = {
  label: "Button",
  icon: "filter",
};
