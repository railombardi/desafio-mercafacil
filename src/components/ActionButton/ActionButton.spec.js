import { shallowMount, createLocalVue } from "@vue/test-utils";
import ActionButton from "./ActionButton.vue";

const localVue = createLocalVue();

let wrapper = shallowMount(ActionButton, {
  localVue,
});

describe("ActionButton.vue", () => {
  it("Snapshot test", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
  describe("[PROPS] =>", () => {
    it("when the prop label is filled the content of the button should change", async () => {
      await wrapper.setProps({ label: "Test label prop" });
      const button = wrapper.find(".button");
      expect(button.text()).toBe("Test label prop");
    });
    it("when the prop icon is filled, a corresponding icon component should appear", async () => {
      await wrapper.setProps({ icon: "filter" });
      const iconName = wrapper.find(".icon").element.tagName.toLowerCase();
      expect(iconName).toContain("filter");
    });
  });
  describe("[EVENTS] =>", () => {
    it("when click button call method of emit click event", () => {
      wrapper.find(".button").trigger("click");
      expect(wrapper.emitted("click")).toHaveLength(1);
    });
  });
  describe("[COMPUTEDS] =>", () => {
    it("when the prop icon is filled the computed iconComponent changes its value", async () => {
      await wrapper.setProps({ icon: "filter" });
      expect(wrapper.vm.iconComponent).toContain("filter-icon");
    });
  });
});
