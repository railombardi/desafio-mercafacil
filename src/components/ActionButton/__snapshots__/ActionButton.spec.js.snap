// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionButton.vue Snapshot test 1`] = `
<label class="wrapper">
  <!----> <button class="button">
    label
  </button>
</label>
`;
